master/pkg/schemas/expconf/log_pattern_config.go

```diff
@@ -4,86 +4,183 @@ import (
+		if names.Contains(*n) {
+			return fmt.Errorf("log_policies have duplicated names %q", *n)
+		}
+		names.Insert(*n)
 	}
-	return out
+
+	*b = LogPoliciesConfig(jsonItems)
+	return nil
 }
 
 // LogPolicyV0 is an action to take if we match against trial logs.
 //
 //go:generate ../gen.sh
 type LogPolicyV0 struct {
-	RawPattern string `json:"pattern"`
-
-	RawAction LogActionV0 `json:"action"`
+	// Legacy log policy doesn't have a name. Legacy log policy will be deprecated.
+	RawName *string `json:"name,omitempty"`
+	// Pattern can be nil. So user can override it to disable the default log polices.
+	RawPattern *string      `json:"pattern,omitempty"`
+	RawAction  *LogActionV0 `json:"action,omitempty"`
 }
 
-// LogActionV0 is a policy to take after matching.
-//
-//go:generate ../gen.sh
-type LogActionV0 struct {
-	RawCancelRetries *LogActionCancelRetriesV0 `union:"type,cancel_retries" json:"-"`
-	RawExcludeNode   *LogActionExcludeNodeV0   `union:"type,exclude_node" json:"-"`
-}
+// LogActionType is the type of an action.
+type LogActionType string
 
-// Merge implements schemas.Mergeable.
-func (s LogActionV0) Merge(other LogActionV0) LogActionV0 {
-	return schemas.UnionMerge(s, other)
-}
+// LogActionType refers to the action user can take when a pattern is detected in the log.
+const (
+	LogActionTypeCancelRetries LogActionType = "cancel_retries"
+	LogActionTypeExcludeNode   LogActionType = "exclude_node"
+)
 
-// MarshalJSON implements the json.Marshaler interface.
-func (s LogActionV0) MarshalJSON() ([]byte, error) {
-	return union.MarshalEx(s, true)
+// LogActionV0 is a policy to take after matching.
+type LogActionV0 struct {
+	Type LogActionType
 }
 
 // UnmarshalJSON implements the json.Unmarshaler interface.
+// It applies a shim to the legacy actions. For example we have a legacy action:
+//
+//	action:
+//	  type: cancel_retries
+//
+// It will become:
+//
+//	action: cancel_retries
 func (s *LogActionV0) UnmarshalJSON(data []byte) error {
-	if err := union.Unmarshal(data, s); err != nil {
-		return err
+	// First, check if we can unmarshal the log policy item as a legacy item.
+	type LegacyAction struct {
+		Type LogActionType `json:"type"`
+	}
+
+	var legacy LegacyAction
+	err := json.Unmarshal(data, &legacy)
+	if err == nil {
+		// Apply shim to bring legacy policy into the current format.
+		switch legacy.Type {
+		case LogActionTypeCancelRetries, LogActionTypeExcludeNode:
+			*s = LogActionV0(legacy)
+		default:
+			return fmt.Errorf("unrecognized legacy action type: %s, data: %q", legacy.Type, string(data))
+		}
+		return nil
 	}
-	type DefaultParser *LogActionV0
-	if err := json.Unmarshal(data, DefaultParser(s)); err != nil {
-		return fmt.Errorf("failed to parse searcher config: %w", err)
+
+	// It is not a legacy item. Try to unmarshal it as a modern item.
+	var lat LogActionType
+	if err := json.Unmarshal(data, &lat); err == nil {
+		switch lat {
+		case LogActionTypeCancelRetries, LogActionTypeExcludeNode:
```

Comment:
if the type isn't an expected one we end up fmt.Errorf'ing a nil err.

Reply:
Add default to switch.